In [6]:
#PIP INSTALLATION
print('---1---')
%pip install pandas
print('---2---')
%pip install seaborn
print('---3---')
%pip install scikit-learn
print('---4---')
%pip install catboost
print('---5---')
%pip install ipywidgets
print('---6---')
%pip install pandas openpyxl
print('---7---')
%pip install bayesian-optimization


---1---
---2---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip



Note: you may need to restart the kernel to use updated packages.
---3---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
---4---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
---5---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
---6---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
---7---



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [7]:
%pip list

Package               Version
--------------------- -----------
asttokens             3.0.0
bayesian-optimization 2.0.3
blinker               1.9.0
catboost              1.2.7
click                 8.1.7
colorama              0.4.6
comm                  0.2.2
contourpy             1.3.1
cycler                0.12.1
debugpy               1.8.12
decorator             5.1.1
et_xmlfile            2.0.0
executing             2.1.0
Flask                 3.1.0
Flask-SQLAlchemy      3.1.1
fonttools             4.55.3
graphviz              0.20.3
greenlet              3.1.1
ipykernel             6.29.5
ipython               8.31.0
ipywidgets            8.1.5
itsdangerous          2.2.0
jedi                  0.19.2
Jinja2                3.1.4
joblib                1.4.2
jupyter_client        8.6.3
jupyter_core          5.7.2
jupyterlab_widgets    3.0.13
kiwisolver            1.4.7
MarkupSafe            3.0.2
matplotlib            3.10.0
matplotlib-inline     0.1.7
nest-asyncio          1.6.0
num

In [8]:
#FROM LIBRARIES
from bayes_opt import BayesianOptimization
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from catboost import CatBoostClassifier, Pool
from IPython.display import FileLink
from sklearn.preprocessing import LabelEncoder
from sklearn import metrics
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, roc_curve, auc, classification_report

#IMPORT AS
import numpy as np
import pandas as pd
import seaborn as sns
import sklearn
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')

In [9]:
file_path = 'Dataset/diabetic_data.csv'
df = pd.read_csv(file_path,keep_default_na=False)
df_clean = df.copy()
df_clean.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [10]:
df_clean.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

In [11]:
print(list(df_clean))

['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']


Data Cleaning 

In [12]:
#REMOVING DUPLICATE DATA'S
df_clean['patient_nbr'].value_counts()

patient_nbr
88785891     40
43140906     28
1660293      23
88227540     23
23199021     23
             ..
11005362      1
98252496      1
1019673       1
13396320      1
175429310     1
Name: count, Length: 71518, dtype: int64

In [13]:
df_clean = df_clean.drop_duplicates(subset=['patient_nbr'])
#END REMOVING DUPLICATE DATA'S

There are 30,248 total of Duplicate Data's which the unique data consist only 71,518

In [14]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71518 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   encounter_id              71518 non-null  int64 
 1   patient_nbr               71518 non-null  int64 
 2   race                      71518 non-null  object
 3   gender                    71518 non-null  object
 4   age                       71518 non-null  object
 5   weight                    71518 non-null  object
 6   admission_type_id         71518 non-null  int64 
 7   discharge_disposition_id  71518 non-null  int64 
 8   admission_source_id       71518 non-null  int64 
 9   time_in_hospital          71518 non-null  int64 
 10  payer_code                71518 non-null  object
 11  medical_specialty         71518 non-null  object
 12  num_lab_procedures        71518 non-null  int64 
 13  num_procedures            71518 non-null  int64 
 14  num_medications           

Value of each drug that most patient uses

-/-/-/-/-/-/-

In [15]:
print(df_clean['metformin'].value_counts())
print("-----------------------------")
print(df_clean['repaglinide'].value_counts())
print("-----------------------------")
print(df_clean['nateglinide'].value_counts())
print("-----------------------------")
print(df_clean['chlorpropamide'].value_counts())

metformin
No        56527
Steady    13718
Up          838
Down        435
Name: count, dtype: int64
-----------------------------
repaglinide
No        70592
Steady      824
Up           73
Down         29
Name: count, dtype: int64
-----------------------------
nateglinide
No        71020
Steady      474
Up           16
Down          8
Name: count, dtype: int64
-----------------------------
chlorpropamide
No        71446
Steady       67
Up            4
Down          1
Name: count, dtype: int64


In [16]:
print(df_clean['glimepiride'].value_counts())
print("-----------------------------")
print(df_clean['acetohexamide'].value_counts())
print("-----------------------------")
print(df_clean['glipizide'].value_counts())
print("-----------------------------")
print(df_clean['glyburide'].value_counts())


glimepiride
No        67773
Steady     3372
Up          235
Down        138
Name: count, dtype: int64
-----------------------------
acetohexamide
No        71517
Steady        1
Name: count, dtype: int64
-----------------------------
glipizide
No        62412
Steady     8150
Up          578
Down        378
Name: count, dtype: int64
-----------------------------
glyburide
No        63664
Steady     6812
Up          621
Down        421
Name: count, dtype: int64


In [17]:
print(df_clean['tolbutamide'].value_counts())
print("-----------------------------")
print(df_clean['pioglitazone'].value_counts())
print("-----------------------------")
print(df_clean['rosiglitazone'].value_counts())
print("-----------------------------")
print(df_clean['acarbose'].value_counts())

tolbutamide
No        71499
Steady       19
Name: count, dtype: int64
-----------------------------
pioglitazone
No        66210
Steady     5047
Up          180
Down         81
Name: count, dtype: int64
-----------------------------
rosiglitazone
No        66817
Steady     4490
Up          136
Down         75
Name: count, dtype: int64
-----------------------------
acarbose
No        71316
Steady      192
Up           10
Name: count, dtype: int64


In [18]:
print(df_clean['miglitol'].value_counts())
print("-----------------------------")
print(df_clean['troglitazone'].value_counts())
print("-----------------------------")
print(df_clean['tolazamide'].value_counts())
print("-----------------------------")
print(df_clean['examide'].value_counts())

miglitol
No        71498
Steady       18
Down          1
Up            1
Name: count, dtype: int64
-----------------------------
troglitazone
No        71515
Steady        3
Name: count, dtype: int64
-----------------------------
tolazamide
No        71488
Steady       30
Name: count, dtype: int64
-----------------------------
examide
No    71518
Name: count, dtype: int64


In [19]:
print(df_clean['citoglipton'].value_counts())
print("-----------------------------")
print(df_clean['insulin'].value_counts())
print("-----------------------------")
print(df_clean['glyburide-metformin'].value_counts())
print("-----------------------------")
print(df_clean['glipizide-metformin'].value_counts())

citoglipton
No    71518
Name: count, dtype: int64
-----------------------------
insulin
No        34921
Steady    22129
Down       7505
Up         6963
Name: count, dtype: int64
-----------------------------
glyburide-metformin
No        71016
Steady      491
Up            7
Down          4
Name: count, dtype: int64
-----------------------------
glipizide-metformin
No        71511
Steady        7
Name: count, dtype: int64


In [20]:
print(df_clean['glimepiride-pioglitazone'].value_counts())
print("-----------------------------")
print(df_clean['metformin-rosiglitazone'].value_counts())
print("-----------------------------")
print(df_clean['metformin-pioglitazone'].value_counts())

glimepiride-pioglitazone
No    71518
Name: count, dtype: int64
-----------------------------
metformin-rosiglitazone
No        71516
Steady        2
Name: count, dtype: int64
-----------------------------
metformin-pioglitazone
No        71517
Steady        1
Name: count, dtype: int64


In [21]:
(df_clean.medical_specialty.value_counts(normalize=True, dropna=False) * 100)


medical_specialty
?                                   48.207444
InternalMedicine                    15.267485
Family/GeneralPractice               7.156240
Emergency/Trauma                     6.243184
Cardiology                           5.964932
                                      ...    
SportsMedicine                       0.001398
Dermatology                          0.001398
Proctology                           0.001398
Surgery-PlasticwithinHeadandNeck     0.001398
Resident                             0.001398
Name: proportion, Length: 71, dtype: float64

Removing features that is not useful for our target readmittied

encounter_id  > Irrelevant 

Patient_nbr > Irrelevant

Weight > Consist of 97% missing value

Payer_code > Irrelevant

Medicial_Speciality > 48% missing value

Repaglinide	> 71K samples not using the drug

Nateglinide	> 71K samples not using the drug

Chlorpropamide > 71K samples not using the drug

Acarbose > 71k samples not using the drug

Miglitol > 71k samples not using the drug

Troglitazone > Only 3 patients using the drug

Tolazamide > Only 39 patients using the drug

Examide	> All patients not using the drug

Citoglipton	> All patients not using the drug

Glyburide_metformin	> 71k samples not using the drug

Glipizide_metfotmin > Only 13 patients using the drug

Glimepiride_pioglitazone > Only 1 patient using the drug

Metformin_rosiglitazone > Only 2 patients using the drug

Metformin_pioglitazone > Only 1 patient using the drug

Acetohexamide > Only 1 patient using the drug

tolbutamide > Only 23 patients using the drug



In [22]:
features_drop_list = ['encounter_id', 'patient_nbr', 'weight', 'payer_code', 'medical_specialty', 'repaglinide', 'nateglinide', 'chlorpropamide', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone','metformin-pioglitazone', 'acetohexamide', 'tolbutamide']
df_clean.drop(features_drop_list, axis=1, inplace=True)

In [23]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71518 entries, 0 to 101765
Data columns (total 29 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   race                      71518 non-null  object
 1   gender                    71518 non-null  object
 2   age                       71518 non-null  object
 3   admission_type_id         71518 non-null  int64 
 4   discharge_disposition_id  71518 non-null  int64 
 5   admission_source_id       71518 non-null  int64 
 6   time_in_hospital          71518 non-null  int64 
 7   num_lab_procedures        71518 non-null  int64 
 8   num_procedures            71518 non-null  int64 
 9   num_medications           71518 non-null  int64 
 10  number_outpatient         71518 non-null  int64 
 11  number_emergency          71518 non-null  int64 
 12  number_inpatient          71518 non-null  int64 
 13  diag_1                    71518 non-null  object
 14  diag_2                    

ICD 9 codes

In [24]:
#start by setting all values containing E or V into 0 (as one category)
df_clean.loc[df_clean['diag_1'].str.contains('V',na=False,case=False), 'diag_1'] = 0
df_clean.loc[df_clean['diag_1'].str.contains('E',na=False,case=False), 'diag_1'] = 0
df_clean.loc[df_clean['diag_2'].str.contains('V',na=False,case=False), 'diag_2'] = 0
df_clean.loc[df_clean['diag_2'].str.contains('E',na=False,case=False), 'diag_2'] = 0
df_clean.loc[df_clean['diag_3'].str.contains('V',na=False,case=False), 'diag_3'] = 0
df_clean.loc[df_clean['diag_3'].str.contains('E',na=False,case=False), 'diag_3'] = 0

#setting all missing values into -1
df_clean['diag_1'] = df_clean['diag_1'].replace('?', -1)
df_clean['diag_2'] = df_clean['diag_2'].replace('?', -1)
df_clean['diag_3'] = df_clean['diag_3'].replace('?', -1)

#No all diag values can be converted into numeric values
df_clean['diag_1'] = df_clean['diag_1'].astype(float)
df_clean['diag_2'] = df_clean['diag_2'].astype(float)
df_clean['diag_3'] = df_clean['diag_3'].astype(float)

In [25]:
# Ensure 'diag_1', 'diag_2', 'diag_3' are numeric
df_clean[['diag_1', 'diag_2', 'diag_3']] = df_clean[['diag_1', 'diag_2', 'diag_3']].apply(pd.to_numeric, errors='coerce')

def categorize_diag(df_clean, col):
    df_clean.loc[(df_clean[col] >= 1) & (df_clean[col] < 140), col] = 1
    df_clean.loc[(df_clean[col] >= 140) & (df_clean[col] < 240), col] = 2
    df_clean.loc[(df_clean[col] >= 240) & (df_clean[col] < 280), col] = 3
    df_clean.loc[(df_clean[col] >= 280) & (df_clean[col] < 290), col] = 4
    df_clean.loc[(df_clean[col] >= 290) & (df_clean[col] < 320), col] = 5
    df_clean.loc[(df_clean[col] >= 320) & (df_clean[col] < 390), col] = 6
    df_clean.loc[(df_clean[col] >= 390) & (df_clean[col] < 460), col] = 7
    df_clean.loc[(df_clean[col] >= 460) & (df_clean[col] < 520), col] = 8
    df_clean.loc[(df_clean[col] >= 520) & (df_clean[col] < 580), col] = 9
    df_clean.loc[(df_clean[col] >= 580) & (df_clean[col] < 630), col] = 10
    df_clean.loc[(df_clean[col] >= 630) & (df_clean[col] < 680), col] = 11
    df_clean.loc[(df_clean[col] >= 680) & (df_clean[col] < 710), col] = 12
    df_clean.loc[(df_clean[col] >= 710) & (df_clean[col] < 740), col] = 13
    df_clean.loc[(df_clean[col] >= 740) & (df_clean[col] < 760), col] = 14
    df_clean.loc[(df_clean[col] >= 760) & (df_clean[col] < 780), col] = 15
    df_clean.loc[(df_clean[col] >= 780) & (df_clean[col] < 800), col] = 16
    df_clean.loc[(df_clean[col] >= 800) & (df_clean[col] < 1000), col] = 17
    df_clean.loc[df_clean[col] == -1, col] = 0  # Handle missing values
    
# Apply function to diagnosis columns
for col in ['diag_1', 'diag_2', 'diag_3']:
    categorize_diag(df_clean, col)

# Print the result
print(df_clean[['diag_1', 'diag_2', 'diag_3']].head())

   diag_1  diag_2  diag_3
0     3.0     0.0     0.0
1     3.0     3.0     3.0
2    11.0     3.0     0.0
3     1.0     3.0     7.0
4     2.0     2.0     3.0


In [26]:
print(df_clean['diag_2'])

0          0.0
1          3.0
2          3.0
3          3.0
4          2.0
          ... 
101754     9.0
101755    10.0
101756    10.0
101758     1.0
101765     9.0
Name: diag_2, Length: 71518, dtype: float64


In [27]:
#check the results
df_clean.diag_2.unique()

array([ 0.,  3.,  2.,  7.,  8.,  4., 17., 13., 10.,  1.,  9., 12.,  6.,
       16.,  5., 11., 14.])

In [28]:
# 1. Basic value counts to see distribution of categories
print("Distribution of diagnosis 1 categories:")
print(df_clean['diag_1'].value_counts().sort_index())


Distribution of diagnosis 1 categories:
diag_1
0.0       939
1.0      1818
2.0      2742
3.0      7685
4.0       657
5.0      1548
6.0       876
7.0     21821
8.0      6736
9.0      6403
10.0     3488
11.0      586
12.0     1789
13.0     4080
14.0       41
16.0     5530
17.0     4779
Name: count, dtype: int64


In [29]:
print("\nDistribution of diagnosis 2 categories:")
print(df_clean['diag_2'].value_counts().sort_index())


Distribution of diagnosis 2 categories:
diag_2
0.0      2094
1.0      1296
2.0      1750
3.0     15482
4.0      2108
5.0      1866
6.0       926
7.0     22206
8.0      6761
9.0      2756
10.0     5180
11.0      353
12.0     2253
13.0     1297
14.0       83
16.0     3249
17.0     1858
Name: count, dtype: int64


In [30]:
print("\nDistribution of diagnosis 3 categories:")
print(df_clean['diag_3'].value_counts().sort_index())


Distribution of diagnosis 3 categories:
diag_3
0.0      4766
1.0      1262
2.0      1262
3.0     19237
4.0      1752
5.0      2164
6.0      1281
7.0     21050
8.0      4466
9.0      2488
10.0     3937
11.0      273
12.0     1532
13.0     1378
14.0       75
16.0     3152
17.0     1443
Name: count, dtype: int64


In [31]:
df_clean['race'] = df_clean['race'].replace('?', 'Other')

In [32]:
df_clean.race.value_counts()

race
Caucasian          53491
AfricanAmerican    12887
Other               3126
Hispanic            1517
Asian                497
Name: count, dtype: int64

In [33]:
print(df_clean['race'].isna())

0         False
1         False
2         False
3         False
4         False
          ...  
101754    False
101755    False
101756    False
101758    False
101765    False
Name: race, Length: 71518, dtype: bool


In [34]:
df_clean.gender.value_counts()

gender
Female             38025
Male               33490
Unknown/Invalid        3
Name: count, dtype: int64

In [35]:
df_clean['gender'] = df_clean['gender'].replace('Unknown/Invalid', 'Female')
df_clean.gender.value_counts()

gender
Female    38028
Male      33490
Name: count, dtype: int64

In [36]:
df_clean.age.value_counts()


age
[70-80)     18210
[60-70)     15960
[50-60)     12466
[80-90)     11589
[40-50)      6878
[30-40)      2699
[90-100)     1900
[20-30)      1127
[10-20)       535
[0-10)        154
Name: count, dtype: int64

In [37]:
for i in range(0, 10):
    age_range = f'[{10*i}-{10*(i+1)})'
    midpoint = i*10 + 5
    df_clean['age'] = df_clean['age'].replace(age_range, midpoint)

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\3540650471.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_clean['age'] = df_clean['age'].replace(age_range, midpoint)


In [38]:
df_clean['age'].value_counts()

age
75    18210
65    15960
55    12466
85    11589
45     6878
35     2699
95     1900
25     1127
15      535
5       154
Name: count, dtype: int64

-- Start Here --

max_glu_serum -> None, Norm, >200, >300


In [39]:
df_clean.max_glu_serum.value_counts()


max_glu_serum
None    68062
Norm     1731
>200      969
>300      756
Name: count, dtype: int64

In [40]:
df['max_glu_serum']=df['max_glu_serum'].replace("None", 0)
df['max_glu_serum']=df['max_glu_serum'].replace("Norm", 1)
df['max_glu_serum']=df['max_glu_serum'].replace(">200", 2)
df['max_glu_serum']=df['max_glu_serum'].replace(">300", 3)
df.max_glu_serum.value_counts()

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\3494085451.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['max_glu_serum']=df['max_glu_serum'].replace(">300", 3)


max_glu_serum
0    96420
1     2597
2     1485
3     1264
Name: count, dtype: int64

In [41]:
df.A1Cresult.value_counts()


A1Cresult
None    84748
>8       8216
Norm     4990
>7       3812
Name: count, dtype: int64

In [42]:
df['A1Cresult']=df['A1Cresult'].replace("None", 0)
df['A1Cresult']=df['A1Cresult'].replace("Norm", 1)
df['A1Cresult']=df['A1Cresult'].replace(">7", 2)
df['A1Cresult']=df['A1Cresult'].replace(">8", 3)


df.A1Cresult.value_counts()

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\720045810.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['A1Cresult']=df['A1Cresult'].replace(">8", 3)


A1Cresult
0    84748
3     8216
1     4990
2     3812
Name: count, dtype: int64

In [43]:
drug_list = ['metformin', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone', 'rosiglitazone', 'insulin']
for i in drug_list:
    df[i] = df[i].replace('No', 0)
    df[i] = df[i].replace('Steady', 2)
    df[i] = df[i].replace('Down', 1)
    df[i] = df[i].replace('Up', 3)

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\1513858272.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[i] = df[i].replace('Up', 3)


In [44]:
#df.metformin.value_counts()
#df.glimepiride.value_counts()
#df.glipizide.value_counts()
#df.glyburide.value_counts()
#df.pioglitazone.value_counts()
#df.rosiglitazone.value_counts()
df.insulin.value_counts()

insulin
0    47383
2    30849
1    12218
3    11316
Name: count, dtype: int64

In [45]:
df.change.value_counts()

change
No    54755
Ch    47011
Name: count, dtype: int64

In [46]:
df['change']=df['change'].replace('No', 0)
df['change']=df['change'].replace('Ch', 1)
df.change.value_counts()

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\746731630.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['change']=df['change'].replace('Ch', 1)


change
0    54755
1    47011
Name: count, dtype: int64

In [47]:
df.diabetesMed.value_counts()

diabetesMed
Yes    78363
No     23403
Name: count, dtype: int64

In [48]:
df['diabetesMed']=df['diabetesMed'].replace('Yes', 1)
df['diabetesMed']=df['diabetesMed'].replace('No', 0)
df.diabetesMed.value_counts()

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\1214750202.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['diabetesMed']=df['diabetesMed'].replace('No', 0)


diabetesMed
1    78363
0    23403
Name: count, dtype: int64

In [49]:
df['readmitted']=df['readmitted'].replace('NO', 0)
df['readmitted']=df['readmitted'].replace('>30', 1)
df['readmitted']=df['readmitted'].replace('<30', 2)
df.readmitted.value_counts()

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_17176\3724688615.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['readmitted']=df['readmitted'].replace('<30', 2)


readmitted
0    54864
1    35545
2    11357
Name: count, dtype: int64

In [50]:
print(list(df_clean))

['race', 'gender', 'age', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'glimepiride', 'glipizide', 'glyburide', 'pioglitazone', 'rosiglitazone', 'insulin', 'change', 'diabetesMed', 'readmitted']


In [51]:
df_clean.head()

race  gender  age  admission_type_id  discharge_disposition_id  \
0        Caucasian  Female    5                  6                        25   
1        Caucasian  Female   15                  1                         1   
2  AfricanAmerican  Female   25                  1                         1   
3        Caucasian    Male   35                  1                         1   
4        Caucasian    Male   45                  1                         1   

   admission_source_id  time_in_hospital  num_lab_procedures  num_procedures  \
0                    1                 1                  41               0   
1                    7                 3                  59               0   
2                    7                 2                  11               5   
3                    7                 2                  44               1   
4                    7                 1                  51               0   

   num_medications  ...  metformin  glimepiride  glipizide  glyburide  \
0                1  ...         No           No         No         No   
1               18  ...         No           No         No         No   
2               13  ...         No           No     Steady         No   
3               16  ...         No           No         No         No   
4                8  ...         No           No     Steady         No   

   pioglitazone  rosiglitazone  insulin change diabetesMed readmitted  
0            No             No       No     No          No         NO  
1            No             No       Up     Ch         Yes        >30  
2            No             No       No     No         Yes         NO  
3            No             No       Up     Ch         Yes         NO  
4            No             No   Steady     Ch         Yes         NO  

[5 rows x 29 columns]

-- Algorithm --

In [52]:
X = df_clean.drop('readmitted', axis=1)  # Features
y = df_clean['readmitted']               # Target variable


In [53]:
cat_features = [col for col in X.columns if X[col].dtype == 'object']

In [54]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [55]:
class CatBoostBayesianOptimizer:
    def __init__(self, X_train, y_train, cat_features, eval_metric='f1', random_state=42):

        self.X_train = X_train
        self.y_train = y_train
        self.cat_features = cat_features
        self.eval_metric = eval_metric
        self.random_state = random_state
        
    def _objective(self, iterations, learning_rate, depth, l2_leaf_reg, 
                  border_count, colsample_bylevel):
        
        try:
            params = {
                'iterations': int(iterations),
                'learning_rate': float(learning_rate),
                'depth': int(depth),
                'l2_leaf_reg': float(l2_leaf_reg),
                'border_count': int(border_count),
                'rsm': float(colsample_bylevel),
                'random_state': self.random_state,
                'verbose': False
            }
            
            # Prepare data
            X = self.X_train.copy()
            X[self.cat_features] = X[self.cat_features].astype(str)
            train_pool = Pool(X, label=self.y_train, cat_features=self.cat_features)
            
            # Initialize and train model
            model = CatBoostClassifier(
                loss_function='MultiClass',
                **params
            )
            
            model.fit(train_pool, verbose=False)
            y_pred = model.predict(X)
            
            # Calculate score
            if self.eval_metric == 'f1':
                score = f1_score(self.y_train, y_pred, average='weighted')
            else:
                score = (y_pred == self.y_train).mean()
                
            return score
            
        except Exception as e:
            print(f"Error in optimization: {str(e)}")
            return 0.0
    
    def optimize(self, n_iter, n_init_points):
        """
        Run Bayesian optimization to find the best hyperparameters.
        """
        # Define parameter bounds with valid CatBoost ranges
        pbounds = {
            'iterations': (10, 100),            # Number of trees
            'learning_rate': (0.01, 1.0),       # Learning rate
            'depth': (1, 10),                    # Tree depth
            'l2_leaf_reg': (0.1, 10.0),            # L2 regularization
            'border_count': (1, 255),         # Must be integer between 1 and 255
            'colsample_bylevel': (0.1, 1.0)    # Feature subsample ratio
        }
        
        optimizer = BayesianOptimization(
            f=self._objective,
            pbounds=pbounds,
            random_state=self.random_state
        )
        
        optimizer.maximize(
            init_points=n_init_points,
            n_iter=n_iter
        )
        
        # Print results
        print("\nBest parameters found:")
        best_params = optimizer.max['params']
        for param, value in best_params.items():
            if param in ['depth', 'border_count', 'iterations']:
                print(f"{param}: {int(value)}")
            else:
                print(f"{param}: {value:.4f}")
        print(f"\nBest {self.eval_metric} score: {optimizer.max['target']:.4f}")
        
        return best_params
    
    def train_final_model(self, best_params):
        X_train = self.X_train.copy()
        X_train[self.cat_features] = X_train[self.cat_features].astype(str)

        train_pool = Pool(X_train, label=self.y_train, cat_features=self.cat_features)

        final_model = CatBoostClassifier(
            loss_function='MultiClass',
            iterations=int(best_params['iterations']),  # Use optimized iterations
            learning_rate=best_params['learning_rate'],
            depth=int(best_params['depth']),
            l2_leaf_reg=best_params['l2_leaf_reg'],
            border_count=int(best_params['border_count']),
            colsample_bylevel=best_params['colsample_bylevel'],
            random_state=self.random_state,
            verbose=100
        )
        final_model.fit(train_pool, verbose=100)
        return final_model

In [56]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

for fold, (train_idx, val_idx) in enumerate(skf.split(X, y)):
    X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
    X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
    
    # Create and run optimizer for this fold
    optimizer = CatBoostBayesianOptimizer(X_train, y_train, cat_features)
    best_params = optimizer.optimize(n_iter=1, n_init_points=1)

    model = optimizer.train_final_model(best_params)

    # Make predictions on the validation set
    y_pred = model.predict(X_val)

    # Calculate accuracy
    accuracy = accuracy_score(y_val, y_pred)
    print(f"Fold {fold+1} - Accuracy: {accuracy:.4f}")
    

|   iter    |  target   | border... | colsam... |   depth   | iterat... | l2_lea... | learni... |
-------------------------------------------------------------------------------------------------
| 1         | 0.5407    | 96.13     | 0.9556    | 7.588     | 63.88     | 1.645     | 0.1644    |
| 2         | 0.566     | 192.4     | 0.8718    | 6.311     | 42.97     | 4.466     | 0.7884    |

Best parameters found:
border_count: 192
colsample_bylevel: 0.8718
depth: 6
iterations: 42
l2_leaf_reg: 4.4656
learning_rate: 0.7884

Best f1 score: 0.5660
0:	learn: 0.8972216	total: 92.2ms	remaining: 3.78s
41:	learn: 0.8048445	total: 3.84s	remaining: 0us
Fold 1 - Accuracy: 0.6156
|   iter    |  target   | border... | colsam... |   depth   | iterat... | l2_lea... | learni... |
-------------------------------------------------------------------------------------------------
| 1         | 0.5449    | 96.13     | 0.9556    | 7.588     | 63.88     | 1.645     | 0.1644    |
| 2         | 0.5725    | 192.4